src/LuaPrinter.ts

```diff
@@ -129,7 +129,27 @@ export class LuaPrinter {
     private sourceFile: string;
 
     public constructor(private options: CompilerOptions, private emitHost: EmitHost, fileName: string) {
-        this.sourceFile = path.basename(fileName);
+        // Determine file name used in source map
+        const fileNameOnly = path.basename(fileName);
+        if (options.outDir) {
+            const fileDir = path.dirname(fileName);
+            const relativeFileDir = path.relative(this.options.rootDir || process.cwd(), fileDir);
+            if (options.sourceRoot) {
+                // When sourceRoot is specified, just use relative path inside rootDir
+                this.sourceFile = path.join(relativeFileDir, fileNameOnly);
+            } else {
+                // Calculate relative path from rootDir to outDir
+                const absoluteOutDir = path.resolve(options.outDir, relativeFileDir);
+                const relativeOutDir = path.relative(absoluteOutDir, fileDir);
+                this.sourceFile = path.join(relativeOutDir, fileNameOnly);
+            }
```

Comment:
I haven't actually tested that code, but can't it be simplified into something like that:
```suggestion
            const relativeFileName = path.relative(program.getCommonSourceDirectory(), fileName);
            if (options.sourceRoot) {
                // When sourceRoot is specified, just use relative path inside rootDir
                this.sourceFile = relativeFileName;
            } else {
                // Calculate relative path from rootDir to outDir
                const outputPath = path.resolve(options.outDir, relativeFileName);
                this.sourceFile = path.relative(path.dirname(outputPath), fileName);
            }
```